### PR TITLE
Fetch WP Assets from specific post.

### PR DESCRIPTION
**Pull request by jorgemd24**
### Changes proposed in this Pull Request:

Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/1780

This PR adds a new endpoint `assets/suggestions` that returns a list of assets for a specific post (page/product/post/etc),  with the below mapping:

Google text asset | Existing post/taxonomy meta
-- | --
Final URL | Permalink
(1) Headline Var1 Max. 15 chars | Post/taxonomy title
(2) Headline Var2 Max. 30 chars | Post/taxonomy title
(3) Long headline Max. 90 chars | Site title: Post/taxonomy title
(4) Short description Max. 60 chars | Post excerpt/taxonomy description
(5) Description Max. 90 chars | Post excerpt/taxonomy description, tagline
(5) Business name Max. 25 chars | Site title
(7) Call-to-action | –
(8) Display URL path | Slug

See original table: pcTzPl-12o-p2#comment-1849

- The Headlines (1 & 2) are grouped as one headline, as the Ads API only use one type which is [HEADLINE](https://github.com/googleads/google-ads-php/blob/151bab9bd37d46543bbeb2c3b0caedf91cb91b6b/src/Google/Ads/GoogleAds/V11/Enums/AssetFieldTypeEnum/AssetFieldType.php#L177).
- The Descriptions (4 & 5) are grouped as one description, as the Ads API only use one type which is [DESCRIPTION](https://github.com/googleads/google-ads-php/blob/151bab9bd37d46543bbeb2c3b0caedf91cb91b6b/src/Google/Ads/GoogleAds/V11/Enums/AssetFieldTypeEnum/AssetFieldType.php#L178).
- When fetching suggestions from WP/WC the field `call_to_action` will be `null` as it is not possible to offer a suggestion, however when we suggest assets used in other campaigns using the Ads API, the field will be filled. 

In addition to the mapping table, the endpoint returns the image assets linked to the post and the website logo.

At the moment we have three types of images:
- Square Marketing images.
- Marketing Images (landscape).
- Logo

The endpoint will return the same images for Square and Landascape images until we decide how to to resize/format the images. See comment: pcTzPl-12o-p2#comment-2090

The images that will return for a post are a combination of:
- Images attached to the post.
- Gallery images
- In case that is a product, images from the product gallery.
- If the page is the Shop page, it will return images of published products. 

### API Response:

```
{
    "headline": [
        "Sample Page"
    ],
    "long_headline": [
        "Wordpress: Sample Page"
    ],
    "description": [
        "I sell clothing!"
    ],
    "logo": [
        "http://localhost:8082/wp-content/uploads/2022/11/cropped-cropped-ocean-3605547_960_720-150x150.jpeg"
    ],
    "final_url": "http://localhost:8082/sample-page/",
    "business_name": "Wordpress",
    "display_url_path": [
        "sample-page"
    ],
    "square_marketing_images": [
        "http://localhost:8082/wp-content/uploads/2022/11/ocean-3605547_960_720-150x150.jpeg",
        ........
    ],
    "marketing_images": [
        "http://localhost:8082/wp-content/uploads/2022/11/Terminator-2-judgement-day-v3.jpeg",
        ..........
    ],
    call_to_action: null
}
```

### Detailed test instructions:

1. Chose any post created in your website. A post can be a product,page,post or any other custom post_type.
2. Get the ID of that post.
3. Make a request to: `assets/suggestions?id=YOUR_ID&type=post`
4. Check that the response matches the mapping table.

### Additional details:
New PRs will include fetching assets for a specific term/category and assets from other campaigns.

### CODE REVIEW

File: src/Proxies/WP.php [hunk 1]
```diff
@@ -196,6 +196,26 @@ public function get_posts( array $args ): array {
...
+	 * @param string           $output Optional. The required return type. One of OBJECT, ARRAY_A, or ARRAY_N, which
+	 *                                 correspond to a WP_Post object, an associative array, or a numeric array,
+	 *                                 respectively. Default OBJECT.
+	 * @param string           $filter Optional. Type of filter to apply. Accepts 'raw', 'edit', 'db',
+	 *                                 or 'display'. Default 'raw'.
+	 * @return WP_Post|array|null Type corresponding to $output on success or null on failure.
+	 *                            When $output is OBJECT, a `WP_Post` instance is returned.
+	 */
+	public function get_post( $post = null, string $output = OBJECT, string $filter = 'raw' ) {
+		return get_post( $post, $output, $filter );
```
Comment (ianlin) [inline src/Proxies/WP.php:216]: 💅  Since we defined the `get_post` method in WP proxy here, we could use `$this->get_post()` for the places where they use `get_post()` directly in this file. Or maybe replace the code where they use `get_post()` directly in the entire code base, but that should be in another PR as it's out of the scope of this PR.

File: src/Ads/AssetSuggestionsService.php [hunk 2]
```diff
@@ -17,13 +19,181 @@
...
+	 *
+	 * @return array All assets for specific post.
+	 * @throws Exception If the Post ID is invalid.
+	 */
+	protected function get_post_assets( int $id ): array {
+		$post = $this->wp->get_post( $id );
+
+		if ( ! $post ) {
+			throw new Exception(
+				/* translators: 1: is a integer representing an unknown Post ID */
```
Comment (ianlin) [inline src/Ads/AssetSuggestionsService.php]: 💅  an integer.

File: src/Ads/AssetSuggestionsService.php [hunk 3]
```diff
@@ -17,13 +19,181 @@
...
+	 * @param WC $wc WC Proxy.
 	 */
-	public function __construct( WP $wp ) {
+	public function __construct( WP $wp, WC $wc ) {
 		$this->wp = $wp;
+		$this->wc = $wc;
+	}
+
+	/**
+	 * Get assets from specific post or term.
```
Comment (ianlin) [inline src/Ads/AssetSuggestionsService.php]: 💅 Maybe it'd be clearer if we rephrase it as `Get WP and other campaigns' assets from the specific post or term.`?

File: src/Ads/AssetSuggestionsService.php [hunk 4]
```diff
@@ -17,13 +19,181 @@
...
+	 * Get Attachmets for specific posts.
+	 *
+	 * @param array $args See WP_Query::parse_query() for all available arguments.
+	 *
+	 * @return array List of attachments
+	 */
+	protected function get_post_attachments( array $args = [] ): array {
+		$defaults = [
+			'post_type'      => 'attachment',
+			'post_mime_type' => 'image',
```
Comment (ianlin) [inline src/Ads/AssetSuggestionsService.php:179]: 💅 We are only getting `image` mime type in this method, maybe rename the method name to `get_post_image_attachments`?

File: src/Ads/AssetSuggestionsService.php [hunk 5]
```diff
@@ -17,13 +19,181 @@
...
+			]
+		);
+
+		if ( $id === wc_get_page_id( 'shop' ) ) {
+			$attachments_ids = array_merge( $attachments_ids, $this->get_shop_attachments() );
+		}
+
+		if ( $post->post_type === 'product' || $post->post_type === 'product_variation' ) {
+			$product         = $this->wc->maybe_get_product( $id );
+			$attachments_ids = array_merge( $attachments_ids, $product->get_gallery_image_ids(), [ $product->get_image_id() ] );
```
Comment (ianlin) [inline src/Ads/AssetSuggestionsService.php]: Just a personal preference [to use spread operator for array over `array_merge`](https://wiki.php.net/rfc/spread_operator_for_array#advantages_over_array_merge), and I think it's slightly cleaner.

```php
$attachments_ids = [
	...$attachments_ids,
	...$product->get_gallery_image_ids(),
	$product->get_image_id(),
];
```

File: src/Ads/AssetSuggestionsService.php [hunk 6]
```diff
@@ -17,13 +19,181 @@
...
+	/**
+	 * Remove empty values from array.
+	 *
+	 * @param array $array A list of strings.
+	 *
+	 * @return array A list of strings without empty strings.
+	 */
+	protected function remove_empty_values( array $array ): array {
+		return array_values( array_filter( $array ) );
+	}
```
Comment (ianlin) [inline src/Ads/AssetSuggestionsService.php]: 💅 This is more like a utility method so might be better to create something that is shareable like `ArrayUtil.php` or a trait. But it's out of the scope of this PR so we can do it maybe in the cooldown for some refactoring.

File: src/Ads/AssetSuggestionsService.php [hunk 7]
```diff
@@ -17,13 +19,181 @@
...
+
+		$gallery_images_urls = get_post_gallery_images( $id );
+		$marketing_images    = array_merge( $this->get_url_attachments_by_ids( $attachments_ids ), $gallery_images_urls );
+		$marketing_images    = array_slice( $marketing_images, 0, self::DEFAULT_MAXIMUM_MARKETING_IMAGES );
+		$long_headline       = get_bloginfo( 'name' ) . ': ' . $post->post_title;
+
+		return [
+			'headline'                => [ $post->post_title ],
+			'long_headline'           => [ $long_headline ],
+			'description'             => $this->remove_empty_values( [ $post->post_excerpt, get_bloginfo( 'description' ) ] ),
```
Comment (ianlin) [inline src/Ads/AssetSuggestionsService.php]: ❔From the PR description:
> The Headlines (1 & 2) are grouped as one headline
The Descriptions (4 & 5) are grouped as one description

I can see there's only one `description` in the response, but still has two headline: `headline` and `long_headline`. Not sure if I understand it correctly, should it only be just `headline`?

**Review by ianlin — APPROVED** [review]
Thanks for adding the API, I've tested locally and I can get the expected results for posts, products, etc. I just left some non-blocking comments and questions.

File: src/Proxies/WP.php (hunk 1 above)
Comment (jorgemd24) [reply src/Proxies/WP.php:216]: Yes, I agree with you 👍  I will leave this for another PR.

File: src/Ads/AssetSuggestionsService.php (hunk 7 above)
Comment (jorgemd24) [reply src/Ads/AssetSuggestionsService.php]: > headline and long_headline. Not sure if I understand it correctly, should it only be just headline?

Google Ads differentiate between, [HEADLINE](https://github.com/googleads/google-ads-php/blob/151bab9bd37d46543bbeb2c3b0caedf91cb91b6b/src/Google/Ads/GoogleAds/V11/Enums/AssetFieldTypeEnum/AssetFieldType.php#L177) and [LONG_HEADLINE](https://github.com/googleads/google-ads-php/blob/151bab9bd37d46543bbeb2c3b0caedf91cb91b6b/src/Google/Ads/GoogleAds/V11/Enums/AssetFieldTypeEnum/AssetFieldType.php#L193) thats why I kept both fields. For descriptions, Google Ads contains only one type.

**Comment by jorgemd24** [comment]
Thanks, @ianlin for reviewing this! I added your suggestions. 👍
